modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ANode.java

```diff
@@ -55,6 +56,13 @@ public Location getLocation() {
         return location;
     }
 
+    /**
+     * Callback to visit a user tree node.
+     */
+    public <Input, Output> Output visit(UserTreeVisitor<Input, Output> userTreeVisitor, Input input) {
+        throw new UnsupportedOperationException("cannot visit user node type [" + getClass().getCanonicalName() + "]");
```

Comment:
Why is this here just to throw an `UnsupportedOp`?

Reply:
The idea of this is the nodes know how to callback the visitor, but the visitor itself can just call visit on any given node without knowing which specific node it is. Take EBinary as an example, you may want to call leftNode.visit(...) and rightNode.visit(...) without knowing what exactly each node is. The exception here is more of a runtime assert that we tried to visit a node that doesn't support the visitor pattern when it probably should.